Fix the ICU time format conversion logic

PopSlime opened the pull request:
Revise the ICU time format conversion logic to support all unquoted literal texts and the `B` and `b` pattern symbols.

Fix #103592

PopSlime committed:
Fix the ICU time format conversion logic

Revise the ICU time format conversion logic to support all unquoted literal texts and the `B` and `b` pattern symbols.

Fix #103592

PopSlime commented:
I'm new to such a big and complex project and not sure how to test these changes, so this is a draft at the moment. Looking forward to help from anyone.

More information can be found in the fixing issue.

mkhamoyan commented:
@PopSlime thanks for submitting the PR. 
Let's add a test case in https://github.com/dotnet/runtime/blob/main/src/libraries/System.Runtime/tests/System.Globalization.Tests/DateTimeFormatInfo/DateTimeFormatInfoLongTimePattern.cs to validate the implemented changes.

mkhamoyan commented:
/azp run runtime-extra-platforms

mkhamoyan commented:
@PopSlime I was thinking something like below for both longtime and shorttime patterns would validate the change. 
```
[Theory]
[InlineData("zh-TW")]
[InlineData("en-US")]
[InlineData("fr-FR")]
public void LongTimePattern_ValidateAMPMDesignators(string cultureName)
{
    var cultureInfo = new CultureInfo(cultureName);
    var date = DateTime.Today + TimeSpan.FromDays(10) + TimeSpan.FromMinutes(10);
    string formattedDateTime = date.ToString("T", cultureInfo);
    
    Assert.True(formattedDateTime.Contains(cultureInfo.DateTimeFormat.AMDesignator));
}
```

This should be consistent for all platforms and devices.
Now we see issue only in some android devices because of different `CLDR` versions.

PopSlime commented:
@mkhamoyan Is it safe to assume those cultures use a 12-hour clock? `fr-FR` uses a 24-hour clock on my side.

mkhamoyan commented:
> @mkhamoyan Is it safe to assume those cultures use a 12-hour clock? `fr-FR` uses a 24-hour clock on my side.

My bad, `fr-FR` shouldn't be there.

mkhamoyan commented:
> How about this?

Looks good to me. 
> By the way, this fails on my PC (Windows) with `mi: 'h:mm:ss'` and `mi-NZ: 'h:mm:ss'`. Should we address this in another issue?

Yes , we should file an issue for that. 
Let's push in the PR and check how it behaves on other platforms.

tarekgh commented on src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.Icu.cs:
```diff
@@ -378,18 +378,15 @@ private static string ConvertIcuTimeFormatString(ReadOnlySpan<char> icuFormatStr
                         break;
                     case 'a': // AM/PM
+                    case 'b': // am, pm, noon, midnight
+                    case 'B': // flexible day periods
```
Should `B` be translated to `tt` instead? I didn't look at specs yet to see what correct translation should be.

PopSlime replied:
`b` is "am, pm, noon, midnight" and `B` is "flexible day periods," both of which should be acceptable to be translated to `tt` (AM/PM) as no other period is present in .NET if I'm not mistaken.

tarekgh replied:
Looking at: https://unicode-org.github.io/icu/userguide/format_parse/datetime/

|Symbol | Meaning | Pattern | Example Output
| - | - | - | -
a | AM or PM | a, aa, or aaaaaaaaaaaa | PM [abbrev]PM [wide]p
| b | am, pm, noon, midnight | b, bb, or bbbbbbbbbbbb | mid.midnightmd
| B | flexible day periods | B, BB, or BBBBBBBBBBBB | at night [abbrev]at night [wide]at night [narrow]

It looks to me `b` and `B` is not directly translated to `AM/PM`. We need to think more how we can support that in the .NET.

PopSlime replied:
True. But for now I think this is an acceptable solution.

tarekgh replied:
I suggest closing this PR till we figure out how we translate these symbols. Thanks @PopSlime for trying to help with that. Let's continue the discussion on the issue.

tarekgh replied:
> True. But for now I think this is an acceptable solution.

I am inclined to agree with that. This may be better than fully ignoring it.

tarekgh commented on src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.Icu.cs:
```diff
@@ -398,6 +395,13 @@ private static string ConvertIcuTimeFormatString(ReadOnlySpan<char> icuFormatStr
                         break;
 
+                    default:
+                        // Treat non-ASCII characters as literals
```
Why you think we need to do that? this can cause a problem in the future if a new translation is introduced and we are blindly copying it to the translated pattern.

PopSlime replied:
Only ASCII letters are used as pattern symbols in the ICU date time format. It should be safe to copy all other symbols as they are just literal texts.

tarekgh replied:
Take `B` as an example. if we have this suggested code before, the pattern will end up having `B` and will be displayed as `B` in the formatted time which will be very ugly.

PopSlime replied:
Take a closer look. Unrecognized ASCII letters are removed from the pattern.

tarekgh replied:
We crossed our replies. please look at my previous reply.

PopSlime replied:
https://www.unicode.org/reports/tr35/tr35-dates.html#date-format-patterns

> Literal text, which is output as-is when formatting, and must closely match when parsing. Literal text can include:
> - Any characters other than A..Z and a..z, including spaces and punctuation.
> - Any text between single vertical quotes ('xxxx'), which may include A..Z and a..z as literal text.
> - Two adjacent single vertical quotes (''), which represent a literal single quote, either inside or outside quoted text.

Actually I need to fix the double adjacent single vertical quotes pattern, which is not covered in the original implementation as well.

tarekgh replied:
Ok, you are right.

PopSlime replied:
Literal single quote and literal backslash conversion fixed.

PopSlime committed:
Clarify literal texts in the conversion logic

PopSlime committed:
Add tests for verifying time patterns

Add tests verifying that all the short and long time patterns either use
a 24-hour clock or have an AM/PM designator.

PopSlime commented:
@dotnet-policy-service agree

PopSlime committed:
Fix literal single quote and literal backslash conversion

tarekgh commented:
> By the way, this fails on my PC (Windows) with mi: 'h:mm:ss' and mi-NZ: 'h:mm:ss'. Should we address this in another issue?
> Yes , we should file an issue for that.

I don't think we can fix that as the patterns are picked from Windows. I suggest you mark your new test methods with the attribute and see if this can help? 

https://github.com/dotnet/runtime/blob/7934e6449229c57f6c2d7ab9ea1688dc82cf95de/src/libraries/System.Runtime/tests/System.Globalization.Tests/DateTimeFormatInfo/DateTimeFormatInfoLongTimePattern.cs#L264

PopSlime commented:
I'll do that, but I also want to see if similar issues are occurring on other platforms, so let's wait for these checks to be done.

tarekgh commented:
>  this fails on my PC (Windows) with mi: 'h:mm:ss' and mi-NZ: 'h:mm:ss'.

I am trying on my machine, and I am not getting the same results you are getting. 

```
mi-NZ .. Maori (New Zealand) .. h:mm tt
mi .. Maori .. h:mm tt
```

running code like:

```C#
            var ci = CultureInfo.GetCultureInfo("mi-NZ");
            Console.WriteLine($"{ci.Name} .. {ci.EnglishName} .. {ci.DateTimeFormat.ShortTimePattern} .. {ci.DateTimeFormat.LongTimePattern}");
            ci = CultureInfo.GetCultureInfo("mi");
            Console.WriteLine($"{ci.Name} .. {ci.EnglishName} .. {ci.DateTimeFormat.ShortTimePattern} .. {ci.DateTimeFormat.LongTimePattern}");
```

What Windows version do you have? I am wondering how you get this result?

PopSlime commented:
It seems that the version of the CLDR data installed on my PC is below 36. The time format for `mi` was fixed in https://github.com/unicode-org/cldr/pull/104.

- Operating system: Windows 10
- Version number: 22H2
- Internal version: 19045.4529

tarekgh commented:
ok, then the attribute https://github.com/dotnet/runtime/blob/7934e6449229c57f6c2d7ab9ea1688dc82cf95de/src/libraries/System.Runtime/tests/System.Globalization.Tests/DateTimeFormatInfo/DateTimeFormatInfoLongTimePattern.cs#L264 is not going to help. You may try using something like https://github.com/dotnet/runtime/blob/2a779abce2a9dd8722dd3786aca1d09311990222/src/libraries/System.Runtime/tests/System.Globalization.Tests/NumberFormatInfo/NumberFormatInfoData.cs#L10 to avoid the failure.

tarekgh commented:
`fr-CA` is failing with the added test:

```
Bad short time pattern for culture fr-CA: 'HH h mm min'

Bad long time pattern for culture fr-CA: 'HH h mm min ss s'
```

tarekgh commented on src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.Icu.cs:
```diff
@@ -364,32 +364,51 @@ private static string ConvertIcuTimeFormatString(ReadOnlySpan<char> icuFormatStr
                 switch (current)
                 {
+                    case '\\':
+                        result[resultPos++] = '\\';
```
I am not sure what exactly you are trying to do. If pattern has `\` should be kept as it is inside the string and not escaping it by adding extra one. 

For parsing two adjacent two single vertical quotes, it should be easy to check when we get the single quote, check the next character, if it is another single quote, store the two quotes as `\'`. i.e. `''` will be converted to `\'` either inside or outside other quotes. This should simplify the changes here.

tarekgh replied:
```C#
                    case '\'':
                        if (i < icuFormatString.Length - 1 && icuFormatString[i + 1] == '\'')
                        {
                            // Double quote, skip the second one
                            result[resultPos++] = '\\';
                            result[resultPos++] = icuFormatString[i++];
                            continue;
                        }

                        result[resultPos++] = icuFormatString[i++];
                        while (i < icuFormatString.Length)
                        {
                            current = icuFormatString[i];
                            if (current == '\'')
                            {
                                if (i < icuFormatString.Length - 1 && icuFormatString[i + 1] == '\'')
                                {
                                    // Double quote, skip the second one
                                    result[resultPos++] = '\\';
                                    result[resultPos++] = icuFormatString[i++];
                                    i++;
                                    continue;
                                }

                                result[resultPos++] = current;
                                break;
                            }

                            result[resultPos++] = current;
                            i++;
                        }
                        break;
```

PopSlime replied:
Backslashes are not used as part of escape sequences in the ICU format, so we need to duplicate them or they will be escaped in the .NET format.

tarekgh replied:
ok, thanks.

Is there an example of any ICU pattern that uses backslash? I am curious to see such case.

PopSlime replied:
There isn't but in case there will be such cases in the future.

PopSlime replied:
> To handle parsing two adjacent single vertical quotes, you can simply check the next character when encountering a single quote. If it is another single quote, store the pair as `\'`. Thus, `''` will be converted to `\'` regardless of whether it appears inside or outside other quotes. This approach should simplify the necessary changes.

Agree. I'll work on that later.

PopSlime commented:
> Looks to me you have something wrong in parsing the single vertical quotes. The patterns should be `HH 'h' mm` and `HH 'h' mm 'min' ss 's'`

This occurs only on the `browser-Release-wasm-Mono` platform, which seems to be using time formats from a source other than ICU and is not related to the conversion logic here. Should we address this issue later?

PopSlime committed:
Refactor the literal quote conversion logic

PopSlime commented:
The incorrect patterns in the browser environment in caused by the following code, which does not handle literal strings and copies them into the pattern as is. Opening another issue to address it.

https://github.com/dotnet/runtime/blob/8ab80b9af668601ebc102760b4c4e60d23ba8d8f/src/mono/browser/runtime/hybrid-globalization/culture-info.ts#L77

PopSlime commented:
> I meant using check like `if (PlatformDetection.ICUVersion.Major >= 68)` .

That'll do, but should we fix the patterns for lower ICU versions despite it is an issue with the CLDR itself? If we do, how should we do that?

PopSlime committed:
Revise the test logic to ignore literal texts and check pattern redundancy

Modify the test logic so that it recognizes literal texts correctly, and
fails if 12-hour and 24-hour clocks are used at the same time.

PopSlime committed:
Revise the test logic to ensure all cultures are tested

tarekgh commented on src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.Icu.cs:
```diff
@@ -364,32 +364,54 @@ private static string ConvertIcuTimeFormatString(ReadOnlySpan<char> icuFormatStr
                 char current = icuFormatString[i];
                 switch (current)
                 {
+                    case '\\':
```
Could you please add a little comment telling ICU formats don't have escaping so we need to escape `\` if we encountered any?

tarekgh commented on src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.Icu.cs:
```diff
@@ -364,32 +364,54 @@ private static string ConvertIcuTimeFormatString(ReadOnlySpan<char> icuFormatStr
+                            return false;
+                        }
+
+                        if (HandleQuoteLiteral(icuFormatString, ref i, result, ref resultPos)) break;
```
please move `break` into its own `{ }` block for readbility.

tarekgh commented on src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.Icu.cs:
```diff
@@ -364,32 +364,54 @@ private static string ConvertIcuTimeFormatString(ReadOnlySpan<char> icuFormatStr
+                                i++;
+                                return true;
+                            }
+                            return false;
```
Maybe easier to have `result[resultPos++] = '\'';` before reurning false and remove this from the calling site.

PopSlime committed:
Add comments to clarify the backslash conversion

PopSlime committed:
Refactor the conversion logic

Simplify some logic and improve readability.

PopSlime commented:
Currently overriding the default time patterns with the following method in my own project, just for reference.

```cs
static string FixTimePattern(string pattern) {
    bool use12Hour = false;
    bool useAMPM = false;
    for (var i = 0; i < pattern.Length; i++) {
        switch (pattern[i]) {
            case 'h': use12Hour = true; break;
            case 't': useAMPM = true; break;
            case '\\': i++; break;
            case '\'':
                for (i++; i < pattern.Length; i++) {
                    var c = pattern[i];
                    if (c == '\'') break;
                    if (c == '\\') i++;
                }
                break;
        }
    }
    if (!use12Hour || useAMPM) return pattern;
    char[] result = pattern.ToCharArray();
    for (var i = 0; i < result.Length; i++) {
        switch (result[i]) {
            case 'h': result[i] = 'H'; break;
            case '\\': i++; break;
            case '\'':
                for (i++; i < result.Length; i++) {
                    var c = result[i];
                    if (c == '\'') break;
                    if (c == '\\') i++;
                }
                break;
        }
    }
    return new string(result);
}
```

This replaces the 12-hour clock with a 24-hour clock if it is used without an AM/PM designator.

tarekgh commented:
> Currently overriding the default time patterns with the following method in my own project, just for reference.

That is fine to do it in the app side. In .NET we have to honor whatever ICU/OS give us even if users think it is wrong. Fixing the time patterns need to raise to the ICU to fix.

tarekgh reviewed (approved):
@PopSlime thanks for providing the fix and working through the test challenges. LGTM. 

We'll wait for a clean CI run and then we can merge.

tarekgh commented:
@PopSlime the tests are still failing. For now, can you exclude running this test with hype globalization? or special case `fr-CA` when encountering it? 

```
Assert.All() Failure: 1 out of 240 items in the collection did not pass.
[147]: Item:  fr-CA
       Error: Bad long time pattern for culture fr-CA: 'HH h mm min ss s'

Assert.All() Failure: 1 out of 240 items in the collection did not pass.
[147]: Item:  fr-CA
       Error: Bad short time pattern for culture fr-CA: 'HH h mm min'
```

PopSlime committed:
Exclude bad ICU patterns from the tests

PopSlime committed:
Exclude the VerifyTimePatterns tests from hybrid globalization on browser

PopSlime committed:
Add missing usings

ilonatommy commented:
> @PopSlime the tests are still failing. For now, can you exclude running this test with hype globalization? or special case `fr-CA` when encountering it? @ilonatommy @mkhamoyan do you have any better suggestion? @PopSlime already logged issue to tracking fixing fr-CA formats.

Current behavior was expected and other `HybridGlobalization` tests are prepared for that. I am looking if we can provide a generic fix. For now this test can be blocked with `PlatformDetection.IsNotHybridGlobalizationOnBrowser`. The fix can get in in a separate PR

mkhamoyan commented:
/azp run runtime-ioslike,runtime-ioslikesimulator,runtime-maccatalyst

mkhamoyan commented on src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.Icu.cs:
```diff
@@ -364,32 +364,58 @@ private static string ConvertIcuTimeFormatString(ReadOnlySpan<char> icuFormatStr
+                        {
+                            break;
+                        }
+                        for (i++; i < icuFormatString.Length; i++)
```
Nit:
```
 i++;
 for (; i < icuFormatString.Length; i++)
```

mkhamoyan reviewed (approved):
Looks good to me. 
Let's wait for ios pipelines and then we can merge.

PopSlime committed:
Improve readability of the for-loops

mkhamoyan commented:
/azp run runtime-ioslike,runtime-ioslikesimulator,runtime-maccatalyst

ilonatommy commented on src/libraries/System.Runtime/tests/System.Globalization.Tests/DateTimeFormatInfo/DateTimeFormatInfoShortTimePattern.cs:
```diff
@@ -255,6 +255,41 @@ public void ShortTimePattern_SetReadOnly_ThrowsInvalidOperationException()
+                {
+                    switch (pattern[i])
+                    {
+                        case 'H': use24Hour = true; break;
```
What about `HH` and `hh`?

PopSlime commented:
/azp run runtime

ilonatommy commented:
/azp run runtime
